// @forbes/fbs-zephr-scripts/dist/zephrProductMap.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.premiumPaywallProducts = exports.clientSubscriptionProducts = exports.unlimitedSubscriptionProducts = exports.zephrProducts = void 0;
exports.zephrProducts = {
    unlimitedSubscription: 'R8W03AS',
    clientUnlimitedAccess: 'R3PNAHU',
    unlimitedAccessForEmployees: 'RSKL92C',
    unlimitedAccessForExecutiveComps: 'RZZFK7H',
    stageProfileAdmin: 'RXR0D16',
    prodProfileAdmin: 'RO8HSE9',
    advisorUnlimitedAccess: 'RA6L55F',
    // Forbes Unlimited Articles - International
    unlimitedAccessForInternational: 'RKPEVDB',
    // Forbes Top Advisor Directory
    unlimitedAccessForTopAdvisorDirectory: 'R0EQIX2',
    marketplace: 'MARKETPLACE',
    press2020: 'PRESS2020',
    vendors2020: 'VENDORS2020',
    wealthAdvisorResource: 'RCL934Y',
    // Forbes Unlimited Access for Advisor Employees
    unlimitedAccessForAdvisorEmployee: 'R880NB6',
    // Forbes Unlimited Access for Forbes School of Business and Technology
    unlimitedAccessForFSBT: 'RVEX5N5',
    // FTC Contribs with no Bertie Access
    unlimitedAccessForFTC: 'RETBQ1H',
    // Forbes Unlimited Articles for Center for Education Reform
    unlimitedAccessForCEF: 'RXCC3NS',
    // Forbes Unlimited Articles for Forbes Asia Key Media Reps/Clients/Agencies
    unlimitedAccessForAKMR: 'RKA20OK',
    // Forbes Unlimited Articles for International Licensees
    unlimitedAccessForIL: 'RZYU584',
    // Forbes.com Unlimited Articles | Scale Up 
    unlimitedAccessForScaleUp: 'BRYEYOUU',
    // Forbes.com Unlimited Articles | Schools
    unlimitedAccessForSchools: 'RNQLU4R',
    // Forbes.com Unlimited Articles for ForbesLive
    unlimitedAccessForForbesLive: 'RWOFY78',
    // Forbes.com Unlimited Articles | Corporate
    unlimitedAccessForCorporate: 'RYBR1SC',
};
exports.unlimitedSubscriptionProducts = [
    exports.zephrProducts.unlimitedSubscription,
    exports.zephrProducts.unlimitedAccessForInternational,
];
exports.clientSubscriptionProducts = [
    exports.zephrProducts.clientUnlimitedAccess,
    exports.zephrProducts.advisorUnlimitedAccess,
];
exports.premiumPaywallProducts = Object.values(exports.zephrProducts);
// Window assignment for external services (and nextjs hack)
window.Zephr = window.Zephr || {};
window.Zephr.utils = window.Zephr.utils || {};
window.Zephr.utils = __assign(__assign({}, window.Zephr.utils), { productMaps: {
        unlimitedSubscriptionProducts: exports.unlimitedSubscriptionProducts,
        clientSubscriptionProducts: exports.clientSubscriptionProducts,
        premiumPaywallProducts: exports.premiumPaywallProducts,
    } });
