// @forbes/fbs-ads/dist/fbs-ads/defaults.js
export var defaultAdConfig = {
    mobile: false,
    ad_unit_path: '7175/fdc.forbes/article/standard/default/standard',
    positions: {
        top: {
            slot_id: 'top',
            sizes: [
                [728, 90],
                [970, 250],
                [970, 90],
                [970, 66],
                [1, 1],
                'fluid'
            ],
            criteoZoneId: 1168047
        },
        rec: {
            slot_id: 'rec',
            sizes: [
                [300, 250],
                [300, 600],
                [1, 1],
                'fluid'
            ],
            criteoZoneId: 1168043
        },
        mobile: {
            slot_id: 'mobile',
            sizes: [
                [300, 50],
                [320, 50],
                [1, 1]
            ],
            params: {
                hero: true
            }
        },
        mobilerec: {
            slot_id: 'mobilerec',
            sizes: [
                [300, 250],
                [300, 50],
                [320, 180],
                [320, 50],
                [1, 1]
            ]
        }
    },
    params: {},
    isEurope: false
};
