// @forbes/fbs-ads/dist/medianet/medianet.service.js
import { loadScript } from '@forbes/fbs-assets-loader';
var MedianetBiddingService = /** @class */ (function () {
    function MedianetBiddingService() {
        this.channel = '';
        this.section = '';
    }
    /**
     * Initialize Medianet Header Bidding
     */
    MedianetBiddingService.prototype.init = function (config) {
        var _a = config.params, _b = _a === void 0 ? {} : _a, _c = _b.channel, channel = _c === void 0 ? [] : _c, _d = _b.section, section = _d === void 0 ? [] : _d;
        var ua = navigator.userAgent;
        var isMobileAndroid = ua.match(/Android/i) && ua.match(/Mobile/i);
        var isMobileWindows = ua.match(/Nokia|NOKIA/i) && ua.match(/Symbian|Windows Phone/i);
        var isMobile = ua.match(/iP(hone|od)/i)
            || ua.match(/iPod/i)
            || ua.match(/BlackBerry/i)
            || isMobileAndroid
            || isMobileWindows;
        this.loadMedianet(!!isMobile);
        this.setChannelSection({
            channel: channel,
            section: section
        });
    };
    MedianetBiddingService.prototype.setChannelSection = function (channelSection) {
        var channel = channelSection.channel, section = channelSection.section;
        this.channel = (channel || []).join(',');
        this.section = (section || []).map(function (sectionName) { return sectionName.split(':')[1]; }).join(',');
        this.applyConfig({ channel: this.channel, section: this.section });
    };
    /**
     * Apply Targeting for Medianet Header Bidding
     */
    MedianetBiddingService.prototype.applyConfig = function (data) {
        if (!data) {
            return;
        }
        this.data = data || this.data;
        window.mnjs = window.mnjs || {};
        window.mnjs.channel = data.channel || this.channel;
        window.mnjs.section = data.section || this.section;
    };
    /**
     * Load medianet script
     */
    MedianetBiddingService.prototype.loadMedianet = function (isMobile) {
        // code from medianet 2022 https://forbesmedia.atlassian.net/browse/ADSS-77
        if (!(window['forbes'] || {}).fastFetchMedianet) { // only load the JS if not loaded earlier
            window.mnjs = window.mnjs || {
                customerId: isMobile ? '8CU243I76' : '8CU4ZAPFY',
                domain: window.location.hostname,
                version: 1.0
            };
            var mnjsEndpoint = 'https://warp.media.net/js/tags/clientag.js?'
                + 'cid=' + window.mnjs.customerId
                + '&dn=' + window.mnjs.domain
                + '&version=' + window.mnjs.version;
            loadScript(mnjsEndpoint);
        }
    };
    return MedianetBiddingService;
}());
export { MedianetBiddingService };
