// @forbes/fbs-components/dist/esm/components/GlobalHeader/GlobalHeaderSubscribeToNewsletters/GlobalHeaderSubscribeToNewsletters.js
import React__default from 'react';
import classnames from '../../../node_modules/classnames/index.js';
import SvgEnvelope from '../../../node_modules/@forbes/fbs-icons/dist/svg/envelope.svg.js';
import { trackingHandler } from '../../../shared/utilities/tracking.js';
import { useFbsHeader } from '../hooks/useFbsHeader.js';
import styles from './GlobalHeaderSubscribeToNewsletters.scss.js';

var defaultTrackingCategory = 'Template Area Interaction';
var defaultTrackingLabel = 'newsletter-subscribe-header';
var loggedInSubscriptionLink = 'https://account.forbes.com/newsletters?utm_source=header&utm_medium=newsletter';
var loggedOutSubscriptionLink = 'https://account.forbes.com?utm_source=header&utm_medium=newsletter';
var GlobalHeaderSubscribeToNewsletters = function (_a) {
    var trackingProps = _a.trackingProps, className = _a.className;
    var clickHandler = function () {
        if (trackingProps) {
            trackingHandler({
                track: trackingProps.track,
                category: (trackingProps === null || trackingProps === void 0 ? void 0 : trackingProps.category) || defaultTrackingCategory,
                action: 'click',
            }, (trackingProps === null || trackingProps === void 0 ? void 0 : trackingProps.label) || defaultTrackingLabel);
        }
    };
    // We only need isLoggedIn from the hook, so we don't need to pass any props to it.
    // Default values are fine.
    var isLoggedIn = useFbsHeader({}).isLoggedIn;
    return (React__default.createElement("a", { className: classnames(styles.container, className, 'newsLetterSubscriptionLink'), href: isLoggedIn ? loggedInSubscriptionLink : loggedOutSubscriptionLink, onClick: clickHandler, target: "_blank", rel: "noreferrer" },
        React__default.createElement("span", { className: styles.emailIcon },
            React__default.createElement(SvgEnvelope, null)),
        React__default.createElement("span", { className: styles.subscribeText }, "Subscribe to newsletters")));
};

export { GlobalHeaderSubscribeToNewsletters as default, defaultTrackingCategory, defaultTrackingLabel, loggedInSubscriptionLink, loggedOutSubscriptionLink };
