// @forbes/fbs-components/dist/cjs/components/GlobalHeader/hooks/useFbsHeader.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var zephrUtils = require('./zephrUtils.js');
var adaptHrefForStaging = require('./adaptHrefForStaging.js');

var useFbsHeader = function (_a) {
    var _b, _c;
    var _d = _a.isProd, isProd = _d === void 0 ? true : _d, _e = _a.currentUrl, currentUrl = _e === void 0 ? '' : _e, // on SS that is "meta?.url || meta?.canonicalUrl || 'https://www.forbes.com/'"
    hideSubscribeButton = _a.hideSubscribeButton;
    var _f = React.useState(false), zephrBrowserDecisionsReady = _f[0], setZephrBrowserDecisionsReady = _f[1];
    var _g = React.useState(false), zephrInitialised = _g[0], setZephrInitialised = _g[1];
    React.useEffect(function () {
        var _a, _b, _c;
        setZephrInitialised(!!((_a = window.Zephr) === null || _a === void 0 ? void 0 : _a.user) || !!((_b = window.Zephr) === null || _b === void 0 ? void 0 : _b.error));
        setZephrBrowserDecisionsReady(!!((_c = window.Zephr) === null || _c === void 0 ? void 0 : _c.outcomes));
        var zephrInitCallback = function () { return setZephrInitialised(true); };
        var browserDecisionsCallback = function () { return setZephrBrowserDecisionsReady(true); };
        document.addEventListener('fbs-zephr-init', zephrInitCallback);
        document.addEventListener('zephr.browserDecisionsFinished', browserDecisionsCallback);
        return function () {
            document.removeEventListener('fbs-zephr-init', zephrInitCallback);
            document.removeEventListener('zephr.browserDecisionsFinished', browserDecisionsCallback);
        };
    }, []);
    var isReady = zephrInitialised && zephrBrowserDecisionsReady;
    if (!isReady) {
        return {};
    }
    var zephrUtils$1 = (_b = window.Zephr) === null || _b === void 0 ? void 0 : _b.utils;
    var user = (_c = window.Zephr) === null || _c === void 0 ? void 0 : _c.user;
    var firstName = (user || {})["first-name"];
    var isLoggedIn = !!user;
    var isForYou = (user === null || user === void 0 ? void 0 : user['hide-forbes-services']) !== 'yes';
    var isListsProfile = (user === null || user === void 0 ? void 0 : user['has-list-profile']) === 'yes';
    var redirect = currentUrl || window.location.href;
    var subscribeUrl = adaptHrefForStaging.adaptHrefForStaging("subscribe/?eventSource=header&redirect=".concat(redirect), isProd);
    var showSubscribeButton = !hideSubscribeButton &&
        !zephrUtils.hasZephrProduct(zephrUtils$1, 'unlimitedSubscriptionProducts') &&
        !zephrUtils.hasZephrProduct(zephrUtils$1, 'clientSubscriptionProducts');
    var isPremium = zephrUtils.hasZephrProduct(zephrUtils$1, 'unlimitedSubscriptionProducts');
    return {
        isLoggedIn: isLoggedIn,
        firstName: firstName,
        isForYou: isForYou,
        isListsProfile: isListsProfile,
        isPremium: isPremium,
        showSubscribeButton: showSubscribeButton,
        subscribeUrl: subscribeUrl,
        showLogin: zephrUtils$1 === null || zephrUtils$1 === void 0 ? void 0 : zephrUtils$1.showLogin,
        logout: zephrUtils$1 === null || zephrUtils$1 === void 0 ? void 0 : zephrUtils$1.logout,
    };
};

exports.useFbsHeader = useFbsHeader;
