// @forbes/fbs-tracking/dist/src/fbs-tracking/google-analytics/google-analytics.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleAnalyticsService = void 0;
var fbs_assets_loader_1 = require("@forbes/fbs-assets-loader");
/**
 * This is a service for tracking to GoogleAnalytics through GTM
 */
var GoogleAnalyticsService = /** @class */ (function () {
    function GoogleAnalyticsService() {
        this.initialized = false;
        this.dataLayer = (window.dataLayer = window.dataLayer || []);
        this.containerId = 'GTM-NMQJM4';
    }
    GoogleAnalyticsService.convertData = function (data) {
        if (data.eventLabel) {
            data['event label'] = data.eventLabel;
        }
        if (data.eventCategory) {
            data['event category'] = data.eventCategory;
        }
        if (data.eventAction) {
            data['event action'] = data.eventAction;
        }
        if (data.eventValue) {
            data['event value'] = data.eventValue;
        }
        delete data.eventLabel;
        delete data.eventCategory;
        delete data.eventAction;
        delete data.eventValue;
    };
    /**
     * This fires a GA Call via GTM
     */
    GoogleAnalyticsService.prototype.track = function (data) {
        if (!data) {
            return;
        }
        GoogleAnalyticsService.convertData(data);
        this.dataLayerPush(data);
        if (!this.initialized) {
            this.init();
        }
    };
    /**
     * This puts necessary GA Variables on the page.
     */
    GoogleAnalyticsService.prototype.init = function () {
        this.initialized = true;
        this.dataLayerPush({
            'gtm.start': new Date().getTime(),
            event: 'gtm.js',
        });
        window.analyticsEvent =
            window.analyticsSocial =
                window.analyticsVPV =
                    window.analyticsClearVPV =
                        window.analyticsForm =
                            function () { };
        (0, fbs_assets_loader_1.loadScript)("https://www.googletagmanager.com/gtm.js?id=".concat(this.containerId), function () {
            var event = new CustomEvent('fbs-gtm-init');
            window.dispatchEvent(event);
        });
    };
    /**
     * Just a wrapper around dataLayer.push
     */
    GoogleAnalyticsService.prototype.dataLayerPush = function (data) {
        this.dataLayer.push(data);
    };
    return GoogleAnalyticsService;
}());
exports.GoogleAnalyticsService = GoogleAnalyticsService;
