// @forbes/fbs-components/dist/cjs/components/GlobalHeader/GlobalHeaderSubscribeToNewsletters/GlobalHeaderSubscribeToNewsletters.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var index = require('../../../node_modules/classnames/index.js');
var envelope = require('../../../node_modules/@forbes/fbs-icons/dist/svg/envelope.svg.js');
var tracking = require('../../../shared/utilities/tracking.js');
var useFbsHeader = require('../hooks/useFbsHeader.js');
var GlobalHeaderSubscribeToNewsletters$1 = require('./GlobalHeaderSubscribeToNewsletters.scss.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var defaultTrackingCategory = 'Template Area Interaction';
var defaultTrackingLabel = 'newsletter-subscribe-header';
var loggedInSubscriptionLink = 'https://account.forbes.com/newsletters?utm_source=header&utm_medium=newsletter';
var loggedOutSubscriptionLink = 'https://account.forbes.com?utm_source=header&utm_medium=newsletter';
var GlobalHeaderSubscribeToNewsletters = function (_a) {
    var trackingProps = _a.trackingProps, className = _a.className;
    var clickHandler = function () {
        if (trackingProps) {
            tracking.trackingHandler({
                track: trackingProps.track,
                category: (trackingProps === null || trackingProps === void 0 ? void 0 : trackingProps.category) || defaultTrackingCategory,
                action: 'click',
            }, (trackingProps === null || trackingProps === void 0 ? void 0 : trackingProps.label) || defaultTrackingLabel);
        }
    };
    // We only need isLoggedIn from the hook, so we don't need to pass any props to it.
    // Default values are fine.
    var isLoggedIn = useFbsHeader.useFbsHeader({}).isLoggedIn;
    return (React__default["default"].createElement("a", { className: index["default"](GlobalHeaderSubscribeToNewsletters$1["default"].container, className, 'newsLetterSubscriptionLink'), href: isLoggedIn ? loggedInSubscriptionLink : loggedOutSubscriptionLink, onClick: clickHandler, target: "_blank", rel: "noreferrer" },
        React__default["default"].createElement("span", { className: GlobalHeaderSubscribeToNewsletters$1["default"].emailIcon },
            React__default["default"].createElement(envelope["default"], null)),
        React__default["default"].createElement("span", { className: GlobalHeaderSubscribeToNewsletters$1["default"].subscribeText }, "Subscribe to newsletters")));
};

exports["default"] = GlobalHeaderSubscribeToNewsletters;
exports.defaultTrackingCategory = defaultTrackingCategory;
exports.defaultTrackingLabel = defaultTrackingLabel;
exports.loggedInSubscriptionLink = loggedInSubscriptionLink;
exports.loggedOutSubscriptionLink = loggedOutSubscriptionLink;
