// @forbes/fbs-components/dist/cjs/components/AdContainer/AdContainer.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../_virtual/_tslib.js');
var React = require('react');
var index = require('../../node_modules/classnames/index.js');
var AdContainer$1 = require('./AdContainer.scss.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

exports.AdContainerTypes = void 0;
(function (AdContainerTypes) {
    AdContainerTypes["default"] = "default";
    AdContainerTypes["headerAd"] = "headerAd";
    AdContainerTypes["inlineAd"] = "inlineAd";
    AdContainerTypes["inlineAdMobile"] = "inlineAdMobile";
    AdContainerTypes["stickyAdMobile"] = "stickyAdMobile";
    AdContainerTypes["rightRail"] = "rightRail";
})(exports.AdContainerTypes || (exports.AdContainerTypes = {}));
function AdContainer(_a) {
    var _b, _c;
    var _ref = _a._ref, children = _a.children, className = _a.className, _d = _a.showLabel, showLabel = _d === void 0 ? true : _d, darkMode = _a.darkMode, type = _a.type, restProps = _tslib.__rest(_a, ["_ref", "children", "className", "showLabel", "darkMode", "type"]);
    var classes = index["default"](AdContainer$1["default"].adContainer, (_b = {}, _b[AdContainer$1["default"].withLabel] = showLabel, _b), (_c = {}, _c[AdContainer$1["default"].darkMode] = darkMode, _c), AdContainer$1["default"][type], className);
    return (React__default["default"].createElement("div", _tslib.__assign({ ref: _ref, className: classes, "aria-hidden": "true", role: "presentation", tabIndex: -1 }, restProps), children));
}

exports["default"] = AdContainer;
