// @forbes/fbs-components/dist/cjs/components/FormElements/Textfield/Textfield.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../_virtual/_tslib.js');
var React = require('react');
var index = require('../../../node_modules/classnames/index.js');
var index$1 = require('../../../shared/components/ShowIf/index.js');
var Textfield$1 = require('./Textfield.scss.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

exports.Sizes = void 0;
(function (Sizes) {
    Sizes["small"] = "small";
    Sizes["medium"] = "medium";
    Sizes["large"] = "large";
})(exports.Sizes || (exports.Sizes = {}));
exports.State = void 0;
(function (State) {
    State["valid"] = "valid";
    State["error"] = "error";
})(exports.State || (exports.State = {}));
function Textfield(_a) {
    var inputProps = _a.inputProps, _b = _a.size, size = _b === void 0 ? exports.Sizes.medium : _b, _c = _a.state, state = _c === void 0 ? exports.State.valid : _c, label = _a.label, helperText = _a.helperText, icon = _a.icon, iconTitle = _a.iconTitle, className = _a.className, _ref = _a._ref, restProps = _tslib.__rest(_a, ["inputProps", "size", "state", "label", "helperText", "icon", "iconTitle", "className", "_ref"]);
    var inputId = inputProps.id;
    return (React__default["default"].createElement("div", _tslib.__assign({ className: index["default"](Textfield$1["default"].iconTextfieldContainer, className), ref: _ref }, restProps),
        React__default["default"].createElement(index$1["default"], { condition: !!label },
            React__default["default"].createElement("div", { className: Textfield$1["default"].labelContainer },
                React__default["default"].createElement("label", { htmlFor: inputId, "aria-label": label, className: Textfield$1["default"].fieldLabel }, label))),
        React__default["default"].createElement("div", { className: index["default"](Textfield$1["default"].iconContainer, Textfield$1["default"][size], Textfield$1["default"][state]), "data-testid": "icon-container" },
            React__default["default"].createElement(index$1["default"], { condition: !!(icon || iconTitle) },
                React__default["default"].createElement("div", { className: Textfield$1["default"].iconBox, "data-testid": "icon-box" },
                    React__default["default"].createElement(index$1["default"], { condition: !!icon }, icon),
                    React__default["default"].createElement(index$1["default"], { condition: !!iconTitle },
                        React__default["default"].createElement("span", { className: Textfield$1["default"].iconText }, iconTitle)))),
            React__default["default"].createElement("input", _tslib.__assign({}, inputProps, { className: index["default"](Textfield$1["default"].field, inputProps.className), id: inputId }))),
        React__default["default"].createElement(index$1["default"], { condition: !!helperText },
            React__default["default"].createElement("div", { className: Textfield$1["default"].helperTextContainer },
                React__default["default"].createElement("span", { className: index["default"](Textfield$1["default"].helperText, Textfield$1["default"][state]) }, helperText)))));
}

exports["default"] = Textfield;
