// @jimdo/growth-tools/dist/components/Experimenter/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Variations = exports.Types = exports.JimdoApp = exports.EventLabels = exports.EventActions = exports.Breakpoints = void 0;
var Types;
exports.Types = Types;
(function (Types) {
  Types["EXPERIMENT_ALL"] = "EXPERIMENT_ALL";
  Types["EXPERIMENT_ALL_INIT"] = "EXPERIMENT_ALL_INIT";
  Types["EXPERIMENT_UPDATE"] = "EXPERIMENT_UPDATE";
  Types["EXPERIMENT_BROADCAST_REQUEST"] = "EXPERIMENT_BROADCAST_REQUEST";
  Types["EXPERIMENT_UPDATE_ALLOCATION"] = "EXPERIMENT_UPDATE_ALLOCATION";
  Types["EXPERIMENT_REMOVE_EXPERIMENT"] = "EXPERIMENT_REMOVE_EXPERIMENT";
})(Types || (exports.Types = Types = {}));
var EventActions;
exports.EventActions = EventActions;
(function (EventActions) {
  EventActions["SEEN"] = "seen";
  EventActions["MISSING_VARIATION"] = "missing_variation";
  EventActions["INVALID_EXPERIMENT_GROUP"] = "invalid_experiment_group";
  EventActions["EXCLUDE"] = "exclude";
  EventActions["ALLOCATION"] = "allocation";
  EventActions["ERROR"] = "error";
})(EventActions || (exports.EventActions = EventActions = {}));
var EventLabels;
exports.EventLabels = EventLabels;
(function (EventLabels) {
  EventLabels["ABTEST"] = "ab_test";
  EventLabels["SUCCESS"] = "success";
  EventLabels["MISSING_VARIATION"] = "missing_variation";
  EventLabels["INVALID_EXPERIMENT_GROUP"] = "invalid_experiment_group";
})(EventLabels || (exports.EventLabels = EventLabels = {}));

// get value of object type
var Variations = {
  FLAG_CONTROL: 'control',
  FLAG_TARGET: 'target',
  FLAG_TARGET_A: 'target-a',
  FLAG_TARGET_B: 'target-b',
  FLAG_TARGET_C: 'target-c',
  FLAG_EXCLUDED: 'excluded'
}; // readonly properties

exports.Variations = Variations;
// union type of readonly properties
// Values taken from https://github.com/Jimdo/ui/blob/master/src/components/constants.ts#L38-L43
var Breakpoints;
exports.Breakpoints = Breakpoints;
(function (Breakpoints) {
  Breakpoints["SM"] = "576px";
  Breakpoints["MD"] = "768px";
  Breakpoints["LG"] = "992px";
  Breakpoints["XL"] = "1200px";
})(Breakpoints || (exports.Breakpoints = Breakpoints = {}));
var JimdoApp;
exports.JimdoApp = JimdoApp;
(function (JimdoApp) {
  JimdoApp["LandingPage"] = "lp";
  JimdoApp["UserAccount"] = "user_account";
  JimdoApp["Signup"] = "signup";
  JimdoApp["LogoTool"] = "logo";
  JimdoApp["Dash"] = "dashboard";
  JimdoApp["CMS"] = "dolphin_cms";
  JimdoApp["CHECKOUT"] = "checkout";
  JimdoApp["UNDEFINED"] = "undefined";
})(JimdoApp || (exports.JimdoApp = JimdoApp = {}));
