// @forbes/fbs-ads/dist/integral/integral.service.js
import { loadScript } from '@forbes/fbs-assets-loader';
// load the IAS PET script on the page
export function initIntegral() {
    loadScript('//cdn.adsafeprotected.com/iasPET.1.js', function () {
        console.warn('integral script already loaded');
    });
}
/**
 * Passes the slot to the function that will make the request to IAS and push it onto the queue.
 * @param adSlot The ad slot to send to IAS.
 */
export function requestIasForSlot(adSlot) {
    var __iasPET = window['__iasPET'] || {};
    __iasPET.queue = __iasPET.queue || [];
    __iasPET.pubId = '930392';
    var sizes = adSlot.getSizes().map(function (size) {
        if (size.getWidth && size.getHeight) {
            return [size.getWidth(), size.getHeight()];
        }
        else {
            return [1, 1];
        }
    });
    // function to run after the IAS request is completed.
    /**
     * Sets the targeting on the page and slot level for IAS
     * @param iasTargeting Targeting string returned from the IAS request
     */
    var iasDataHandler = function (iasTargeting) {
        if (iasTargeting) {
            var googleTag_1 = window['googletag'] = window['googletag'] || {
                cmd: []
            };
            var parsedTargeting_1 = JSON.parse(iasTargeting);
            Object.keys(parsedTargeting_1).map(function (key) {
                var iasTargetingValue = parsedTargeting_1[key];
                if (key === 'slots') {
                    Object.keys(iasTargetingValue).forEach(function (otherKey) {
                        adSlot.setTargeting('iasId', iasTargetingValue[otherKey].id);
                    });
                }
                else if (key === 'fr') {
                    googleTag_1.pubads().setTargeting(key, parsedTargeting_1[key]);
                }
                else {
                    Object.keys(iasTargetingValue).forEach(function (otherKey) {
                        googleTag_1.pubads().setTargeting(otherKey, iasTargetingValue[otherKey]);
                    });
                }
            });
        }
    };
    var iasPETSlots = [{
            adSlotId: adSlot.getSlotElementId(),
            adUnitPath: adSlot.getAdUnitPath(),
            // size: can either be a single size (for example, [728, 90])
            // or an array of sizes (for example, [[728, 90], [970, 90]])
            size: sizes
        }];
    // make the request to PET
    __iasPET.queue.push({
        adSlots: iasPETSlots,
        dataHandler: iasDataHandler
    });
}
