// @forbes/fbs-components/dist/cjs/components/deprecated/Subnav/useSubNavItemEvents.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

var useSubNavItemEvents = function (_a) {
    var subMenuData = _a.subMenuData, isOpen = _a.isOpen, setOpen = _a.setOpen, clickEvent = _a.clickEvent, closeAll = _a.closeAll;
    var _b = React.useState(false), isTouchDevice = _b[0], setIsTouchDevice = _b[1];
    var hasSubMenu = !!subMenuData;
    var open = function () { return hasSubMenu && setOpen(true); };
    var close = function () { return hasSubMenu && setOpen(false); };
    var onTouchStart = function () {
        setIsTouchDevice(true);
    };
    var onTouchEnd = function (e) {
        e.stopPropagation();
        if (isOpen)
            close();
        else
            open();
    };
    var onClick = function (e) {
        if (isTouchDevice && hasSubMenu) {
            e.preventDefault(); // do not follow the link
        }
        if (isTouchDevice && !hasSubMenu) {
            closeAll();
        }
        e.stopPropagation();
        clickEvent();
    };
    var onFocus = function () {
        if (isTouchDevice)
            return;
        open();
    };
    var itemHoverProps = isTouchDevice
        ? {}
        : {
            onMouseOver: open,
            onMouseOut: close,
        };
    var linkProps = {
        onTouchStart: onTouchStart,
        onTouchEnd: onTouchEnd,
        onClick: onClick,
        onFocus: onFocus,
    };
    return { hasSubMenu: hasSubMenu, itemHoverProps: itemHoverProps, linkProps: linkProps };
};

exports.useSubNavItemEvents = useSubNavItemEvents;
