// @forbes/fbs-ads/dist/criteo/criteo-bidding.service.js
import { loadScript } from '@forbes/fbs-assets-loader';
var CriteoBiddingService = /** @class */ (function () {
    function CriteoBiddingService() {
    }
    CriteoBiddingService.prototype.init = function () {
        return this.loadCriteo();
    };
    CriteoBiddingService.prototype.fetchBids = function (slots) {
        var _this = this;
        try {
            this.criteo.events.push(function () {
                _this.criteo.RequestBidsOnGoogleTagSlots({
                    networkId: 7525,
                    placements: slots,
                    callback: function () {
                        _this.criteo.SetDFPKeyValueTargeting();
                    },
                });
            });
        }
        catch (error) {
            console.error(error);
            return;
        }
    };
    CriteoBiddingService.prototype.loadCriteo = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            loadScript('https://static.criteo.net/js/ld/publishertag.js', function () {
                _this.criteo = window['Criteo'] = window['Criteo'] || {};
                _this.criteo.events = _this.criteo.events || [];
                resolve();
            });
        });
    };
    return CriteoBiddingService;
}());
export { CriteoBiddingService };
