// @jimdo/growth-tools/dist/helpers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isTargetVariation = exports.isTargetCVariation = exports.isTargetBVariation = exports.isTargetAVariation = exports.isExcludedVariation = exports.isControlVariation = exports.getExperimentGroupById = void 0;
var _components = require("../components");
var getExperimentGroupById = function getExperimentGroupById(allocations, experimentId) {
  var _find;
  return (_find = (allocations || []).find(function (allocation) {
    return allocation.experimentIdentifier === experimentId;
  })) === null || _find === void 0 ? void 0 : _find.group;
};
exports.getExperimentGroupById = getExperimentGroupById;
var isTargetVariation = function isTargetVariation(variation) {
  return variation === _components.Variations.FLAG_TARGET;
};
exports.isTargetVariation = isTargetVariation;
var isTargetAVariation = function isTargetAVariation(variation) {
  return variation === _components.Variations.FLAG_TARGET_A;
};
exports.isTargetAVariation = isTargetAVariation;
var isTargetBVariation = function isTargetBVariation(variation) {
  return variation === _components.Variations.FLAG_TARGET_B;
};
exports.isTargetBVariation = isTargetBVariation;
var isTargetCVariation = function isTargetCVariation(variation) {
  return variation === _components.Variations.FLAG_TARGET_C;
};
exports.isTargetCVariation = isTargetCVariation;
var isControlVariation = function isControlVariation(variation) {
  return variation === _components.Variations.FLAG_CONTROL;
};
exports.isControlVariation = isControlVariation;
var isExcludedVariation = function isExcludedVariation(variation) {
  return variation === _components.Variations.FLAG_EXCLUDED;
};
exports.isExcludedVariation = isExcludedVariation;
