// @forbes/fbs-components/dist/esm/components/GlobalHeader/hooks/useFbsHeader.js
import { useState, useEffect } from 'react';
import { hasZephrProduct } from './zephrUtils.js';
import { adaptHrefForStaging } from './adaptHrefForStaging.js';

var useFbsHeader = function (_a) {
    var _b, _c;
    var _d = _a.isProd, isProd = _d === void 0 ? true : _d, _e = _a.currentUrl, currentUrl = _e === void 0 ? '' : _e, // on SS that is "meta?.url || meta?.canonicalUrl || 'https://www.forbes.com/'"
    hideSubscribeButton = _a.hideSubscribeButton;
    var _f = useState(false), zephrBrowserDecisionsReady = _f[0], setZephrBrowserDecisionsReady = _f[1];
    var _g = useState(false), zephrInitialised = _g[0], setZephrInitialised = _g[1];
    useEffect(function () {
        var _a, _b, _c;
        setZephrInitialised(!!((_a = window.Zephr) === null || _a === void 0 ? void 0 : _a.user) || !!((_b = window.Zephr) === null || _b === void 0 ? void 0 : _b.error));
        setZephrBrowserDecisionsReady(!!((_c = window.Zephr) === null || _c === void 0 ? void 0 : _c.outcomes));
        var zephrInitCallback = function () { return setZephrInitialised(true); };
        var browserDecisionsCallback = function () { return setZephrBrowserDecisionsReady(true); };
        document.addEventListener('fbs-zephr-init', zephrInitCallback);
        document.addEventListener('zephr.browserDecisionsFinished', browserDecisionsCallback);
        return function () {
            document.removeEventListener('fbs-zephr-init', zephrInitCallback);
            document.removeEventListener('zephr.browserDecisionsFinished', browserDecisionsCallback);
        };
    }, []);
    var isReady = zephrInitialised && zephrBrowserDecisionsReady;
    if (!isReady) {
        return {};
    }
    var zephrUtils = (_b = window.Zephr) === null || _b === void 0 ? void 0 : _b.utils;
    var user = (_c = window.Zephr) === null || _c === void 0 ? void 0 : _c.user;
    var firstName = (user || {})["first-name"];
    var isLoggedIn = !!user;
    var isForYou = (user === null || user === void 0 ? void 0 : user['hide-forbes-services']) !== 'yes';
    var isListsProfile = (user === null || user === void 0 ? void 0 : user['has-list-profile']) === 'yes';
    var redirect = currentUrl || window.location.href;
    var subscribeUrl = adaptHrefForStaging("subscribe/?eventSource=header&redirect=".concat(redirect), isProd);
    var showSubscribeButton = !hideSubscribeButton &&
        !hasZephrProduct(zephrUtils, 'unlimitedSubscriptionProducts') &&
        !hasZephrProduct(zephrUtils, 'clientSubscriptionProducts');
    var isPremium = hasZephrProduct(zephrUtils, 'unlimitedSubscriptionProducts');
    return {
        isLoggedIn: isLoggedIn,
        firstName: firstName,
        isForYou: isForYou,
        isListsProfile: isListsProfile,
        isPremium: isPremium,
        showSubscribeButton: showSubscribeButton,
        subscribeUrl: subscribeUrl,
        showLogin: zephrUtils === null || zephrUtils === void 0 ? void 0 : zephrUtils.showLogin,
        logout: zephrUtils === null || zephrUtils === void 0 ? void 0 : zephrUtils.logout,
    };
};

export { useFbsHeader };
