// @jimdo/growth-tools/dist/utilities/hasCookiePermission/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PerformanceCookies = exports.NecessaryCookies = exports.MarketingCookies = exports.FunctionalCookies = void 0;
var NecessaryCookies;
exports.NecessaryCookies = NecessaryCookies;
(function (NecessaryCookies) {
  NecessaryCookies["CKIES_NECESSARY"] = "ckies_necessary";
})(NecessaryCookies || (exports.NecessaryCookies = NecessaryCookies = {}));
var FunctionalCookies;
exports.FunctionalCookies = FunctionalCookies;
(function (FunctionalCookies) {
  FunctionalCookies["CKIES_FUNCTIONAL"] = "ckies_functional";
})(FunctionalCookies || (exports.FunctionalCookies = FunctionalCookies = {}));
var PerformanceCookies;
exports.PerformanceCookies = PerformanceCookies;
(function (PerformanceCookies) {
  PerformanceCookies["CKIES_PERFORMANCE"] = "ckies_performance";
})(PerformanceCookies || (exports.PerformanceCookies = PerformanceCookies = {}));
var MarketingCookies;
exports.MarketingCookies = MarketingCookies;
(function (MarketingCookies) {
  MarketingCookies["CKIES_MARKETING"] = "ckies_marketing";
})(MarketingCookies || (exports.MarketingCookies = MarketingCookies = {}));
