// @jimdo/ui/dist/es/components/Buttons/SocialButtons/buttonSocialConnect/ButtonSocial.js
import { Icon } from '../../../DataDisplay/Icon/Icon.js';
import { iconCheckmarkCircle, iconMail, iconAppleSign, iconGoogleSign, iconFacebookSign } from '../../../DataDisplay/Icon/icons.js';
import classNames from 'classnames';
import React__default, { forwardRef } from 'react';
import styles from './ButtonSocial.module.scss.js';
const ButtonSocial = forwardRef((props, ref) => {
  const {
    social,
    disabled,
    connected,
    className,
    noGutter,
    ...otherProps
  } = props;
  const classes = classNames(styles.buttonSocialConnect, {
    [styles.buttonSocialConnectFacebook]: social === 'fb',
    [styles.buttonSocialConnectGoogle]: social === 'google',
    [styles.buttonSocialConnectApple]: social === 'apple',
    [styles.buttonSocialConnectEmail]: social === 'email',
    [styles.buttonSocialConnectConnected]: connected,
    [styles.noGutter]: noGutter
  }, className);
  const iconClasses = classNames(styles.buttonSocialConnectIcon, {
    [styles.buttonSocialConnectIconFacebook]: social === 'fb',
    [styles.buttonSocialConnectIconGoogle]: social === 'google',
    [styles.buttonSocialConnectIconApple]: social === 'apple',
    [styles.buttonSocialConnectIconEmail]: social === 'email'
  });
  const getIcon = () => {
    if (connected) {
      return iconCheckmarkCircle;
    }
    switch (social) {
      case 'fb':
        return iconFacebookSign;
      case 'google':
        return iconGoogleSign;
      case 'apple':
        return iconAppleSign;
      case 'email':
        return iconMail;
    }
  };
  const valueForDisabled = disabled || Boolean(props.connected);
  if ('href' in otherProps) {
    return (
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      React__default.createElement("a", {
        ...otherProps,
        ref: ref,
        className: classes,
        href: !valueForDisabled ? props.href : undefined
      }, React__default.createElement(Icon, {
        icon: getIcon(),
        className: iconClasses,
        "aria-hidden": "true"
      }), React__default.createElement("span", {
        className: styles.buttonSocialConnectText
      }, props.children))
    );
  }
  return React__default.createElement("button", {
    ...otherProps,
    ref: ref,
    className: classes,
    disabled: valueForDisabled
  }, React__default.createElement(Icon, {
    icon: getIcon(),
    className: iconClasses,
    "aria-hidden": "true"
  }), React__default.createElement("span", {
    className: styles.buttonSocialConnectText
  }, props.children));
});
export { ButtonSocial };
